### PR TITLE
Line ending translation

### PR DESCRIPTION
**Pull request by Riebart**
Adds a new property to `ShellProfile`s, as well as the settings pages, that determines how to translate line endings (Convert all to CRLF, Convert all to LF, and Do not modify) on paste. Conversion is done through regular expression replacement.

Also adds a new command: Paste without newlines, which uses the regular expression to replace \r and \n characters with empty strings. Useful for handling pasting copied multi-line commands that may have had a newline introduced into them through soft wrapping.

Addresses #110

### CODE REVIEW

**Comment by Riebart** [comment]
@skipmeister123 Maybe? Depends on whether that issue is line-ending related, or related to other bugs where newlines appear after each typed character (#97). Notably, I can no longer reproduce the latter, and this should fix anything related to the former.

It can also be used as a workaround for #169 in some cases.

**Comment by skipmeister123** [comment]
I just pulled down the latest code and tested it.  Looks like it is still broken...

--
Hans Kokx
On 1/12/2019 4:39:16 PM, Mike <notifications@github.com> wrote:
@skipmeister123 [https://github.com/skipmeister123] Maybe? Depends on whether that issue is line-ending related, or related to other bugs where newlines appear after each typed character (#97 [https://github.com/felixse/FluentTerminal/issues/97]). Notably, I can no longer reproduce the latter, and this should fix anything related to the former.
It can also be used as a workaround for #169 [https://github.com/felixse/FluentTerminal/issues/169] in some cases.

**Comment by Riebart** [comment]
I'm without my dev environment for a bit, can you confirm that when you paste into other WSL processes, like nano and bash, that the line endings are coming out correctly.

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml [hunk 1]
```diff
@@ -238,7 +238,7 @@
                                     Text="{x:Bind Arguments, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
                                 <ComboBox
                                     Width="150"
-                                    Margin="0,24,0,0"
+                                    Margin="0,24,0,16"
```
Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: This seems to be an accidental/unnecessary/unrelated change?

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml [hunk 2]
```diff
@@ -264,6 +264,19 @@
                                     <HyperlinkButton Command="{x:Bind AddKeyboardShortcutCommand}" Content="Add" />
                                 </StackPanel>
                                 <views:KeyBindingsView ShowCommandName="False" ViewModel="{x:Bind KeyBindings}" />
+                                <TextBlock Margin="0,0,0,8" Text="Line ending translation on paste" />
```
Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: I would expect the margins to be either `0,24,0,8` or `0,36,0,8`

```suggestion
                                <TextBlock Margin="0,36,0,8" Text="Line ending translation on paste" />
```

Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: In spirit of the other option titles in the shell profile configuration I would rephrase the text as ⬇️

If you update this, don't forget to update it [here](https://github.com/felixse/FluentTerminal/blob/4349d3ad44169f4de38365a3fc00568f47edbeaa/FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml#L324) as well.

```suggestion
                                <TextBlock Margin="0,0,0,8" Text="Line ending translation" />
```

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml [hunk 3]
```diff
@@ -264,6 +264,19 @@
...
+                                <RadioButton
+                                    Content="Do not modify line endings on paste"
+                                    GroupName="LineEndingStyle"
+                                    IsChecked="{x:Bind LineEndingStyle, Mode=TwoWay, Converter={StaticResource LineEndingStyleToRadioButtonConverter}, ConverterParameter='DoNotModify'}" />
+                                <RadioButton
+                                    Content="Convert line endings to CR+LF on paste"
+                                    GroupName="LineEndingStyle"
+                                    IsChecked="{x:Bind LineEndingStyle, Mode=TwoWay, Converter={StaticResource LineEndingStyleToRadioButtonConverter}, ConverterParameter='ToCRLF'}" />
+                                <RadioButton
+                                    Content="Convert line endings to LF on paste"
```
Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: I would personally go with shorter descriptions here, e.g.: "Do not modify", "Convert to CRLF", and "Convert to LF". Although it is personal preference, you seem to have done the same in the preview of the settings.

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml (hunk 2 above)
Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: Since the "Keyboard shortcut" setting can grow in size it feels awkward to me that this setting is placed below that. Personally I would move this setting above the "Keyboard shortcut" setting.

If you update this, don't forget to update it [here](https://github.com/felixse/FluentTerminal/blob/4349d3ad44169f4de38365a3fc00568f47edbeaa/FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml#L324) as well.

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml [hunk 4]
```diff
@@ -307,6 +320,9 @@
                                         ViewModel="{x:Bind KeyBindings}" />
                                 </controls:HeaderedContentControl>
 
+
+                                <TextBlock Margin="0,0,0,8" Text="Line ending translation on paste" />
```
Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: I would expect the margins to be either `0,26,0,8`

```suggestion
                                <TextBlock Margin="0,26,0,8" Text="Line ending translation on paste" />
```

Comment (ericcornelissen) [inline FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: The way this is written is inconsistent with the other options, I think it should be more like this:

```suggestion
                                <controls:HeaderedContentControl
                                    Margin="0,36,0,0"
                                    HorizontalContentAlignment="Stretch"
                                    Header="Line ending translation on paste">
                                    <ContentControl
                                        Margin="0,6,0,0"
                                        Content="{x:Bind LineEndingStyle, Mode=OneWay, Converter={StaticResource EnumToDescriptionConverter}}"
                                        ContentTemplateSelector="{StaticResource TerminalThemeTemplateSelector}" />
                                </controls:HeaderedContentControl>
```

File: FluentTerminal.App.ViewModels/TerminalViewModel.cs [hunk 5]
```diff
@@ -341,11 +341,22 @@ private async void OnKeyboardCommandReceived(object sender, string command)
...
+                string content = await _clipboardService.GetText().ConfigureAwait(true);
                 if (content != null)
                 {
+                    // Use the shell profile line-ending translation method to translate the line endings of the pasted buffer
+                    // or delete all of the instances of the newline pattern by replacing with the empty string.
+                    if (command == nameof(Command.Paste))
+                    {
+                        content = _shellProfile.TranslateLineEndings(content);
+                    }
+                    else
```
Comment (ericcornelissen) [inline FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Perhaps make more explicit for clarity (don't know if the maintainers are a fan of this though)?

```suggestion
                    else if (command == nameof(Command.PasteWithoutNewlines))
```

File: FluentTerminal.App.ViewModels/TerminalViewModel.cs [hunk 6]
```diff
@@ -341,11 +341,22 @@ private async void OnKeyboardCommandReceived(object sender, string command)
...
             }
-            else if (command == nameof(Command.Paste))
+            else if ((command == nameof(Command.Paste)) || (command == nameof(Command.PasteWithoutNewlines)))
             {
-                var content = await _clipboardService.GetText().ConfigureAwait(true);
+                string content = await _clipboardService.GetText().ConfigureAwait(true);
                 if (content != null)
                 {
+                    // Use the shell profile line-ending translation method to translate the line endings of the pasted buffer
+                    // or delete all of the instances of the newline pattern by replacing with the empty string.
```
Comment (ericcornelissen) [inline FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Maybe inline the comments inside the if-statement? To be honest I don't think they're really even necessary...

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml (hunk 2 above)
Comment (Riebart) [reply FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: Changed to `Line ending translation mode`

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml (hunk 3 above)
Comment (Riebart) [reply FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: The idea is to keep descriptions short when viewing a config, but descriptive when editing the configuration, to allow people to understand the functionality when they go to edit a profile. Leaving it as is for now.

File: FluentTerminal.App.ViewModels/TerminalViewModel.cs (hunk 5 above)
Comment (Riebart) [reply FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Resolved by converting that section to a `switch() { }`.

File: FluentTerminal.App.ViewModels/TerminalViewModel.cs (hunk 6 above)
Comment (Riebart) [reply FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Resolved by converting that section to a `switch() { }`.

**Comment by Riebart** [comment]
Accepted most of Eric's suggestions. See most recent commit.

File: FluentTerminal.App.ViewModels/TerminalViewModel.cs (hunk 5 above)
Comment (ericcornelissen) [reply FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Great solution 👍

Stylistically I'm unsure whether the braces in `case: { }` are desired, I've personally never seen it and as far as I can tell it's not used anywhere else in the project. That said, don't feel the need to change it unless the maintainer(s) say something about it 😄

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml [hunk 7]
```diff
@@ -280,7 +293,7 @@
                                     Header="Arguments"
                                     Text="{x:Bind Arguments, Mode=OneWay}" />
                                 <controls:HeaderedContentControl
-                                    Margin="0,36,0,0"
+                                    Margin="0,36,0,48"
```
[the remark this answers is not in the document]
Comment (ericcornelissen) [reply FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: Can you revert this change? It seems to be causing this gap ⬇️

gap preview

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml (hunk 1 above)
Comment (ericcornelissen) [reply FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: Can you look at this change again? It seems to be causing this gap ⬇️

gap preview

File: FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml (hunk 3 above)
Comment (ericcornelissen) [reply FluentTerminal.App/Views/SettingsPages/ShellProfileSettings.xaml]: Hmm, that is a good motivation 🤔 Thinking back, I think my main motivation for commenting was that you're repeating "line endings ... on paste" unnecessarily, what about changing it to something like ⬇️?

* * *

Line endings translation on paste
- [ ] Do not modify
- [ ] Convert to CR+LF
- [ ] Convert to LF

* * *

Or in case you want to keep "line endings" for clarity ⬇️:

* * *

Line endings translation on paste
- [ ] Do not modify line endings
- [ ] Convert line endings to CR+LF
- [ ] Convert line endings to LF

File: FluentTerminal.App.ViewModels/TerminalViewModel.cs (hunk 5 above)
Comment (Riebart) [reply FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Stylistically, that's a hold over from ANSI style bracing of C++, and is (at least, in C++), required if you ever define a local scope variable inside a switch case like we do for the two Paste commands). I'm not married to it, but trained myself over years of using `astyle -j --style ansi`.

Comment (ericcornelissen) [reply FluentTerminal.App.ViewModels/TerminalViewModel.cs]: Makes sense, didn't know that. Thanks for the explanation 🙂

**Comment by Riebart** [comment]
Aside from the typo in the commit message, we should be good to merge I think.

**Comment by felixse** [comment]
Great work 👍 
I pushed some changes, mostly about the way the radio button binding is implemented. I know the way I implemented it is a bit verbose, but I had my fair share of troubles using the converter approach in UWP. Plan is to wrap this in a nice little control in the future to keep the ViewModels clean.
